clang/test/SemaCXX/kernel-function-type.cpp

```diff
@@ -0,0 +1,28 @@
+// RUN: %clang_cc1 -fsycl -fsycl-is-device -fsyntax-only -verify %s
+// expected-no-diagnostics
+struct b{};
```

Comment:
SemaCXX is probably the wrong place for this...  This is a minimization that we did, so I'll do my best to explain what happened.

Basically, the kernel_single_task call was emitted as an OpenCL kernel.  Our call to getFullyQualifiedType caused a now-2nd instantiation of zip_iterator<b,b> (first was on line 16 during phase1).

Then, the call to foo  in main causes 'foo' to be instantiated.  The best match for zip_iterator<j,d> is actually now the one instantiated by the kernel_single_task, which is now different from the one made in phase 1.

Thus is_same<zip_iterator<b,b>, zip_iterator<j,d>>::value is not true, despite both being zip_iterator<b,b>.

This is because it is ACTUALLY:

is_same<zip_iterator<b,b> (as instantiated during phase 1 of line 16), zip_iterator<b,b>(instantiated by getFullyQualifiedName)>.

Reply:
This is the test case (long) that is failing. In this test the assert shouldn't go off, as it is supposed to test two same types. But in the current compiler (without this patch), it is failing in the assert.
#include <CL/sycl.hpp>

#include <dpstd/iterators.h>

namespace sycl = cl::sycl;

namespace dpct {
  namespace internal {
    // struct for checking if iterator is heterogeneous or not
    template <typename Iter, typename Void = void> // for non-heterogeneous iterators
            struct is_hetero_iterator
      : std::false_type
    { };

    template <typename Iter> // for heterogeneous iterators
    struct is_hetero_iterator<Iter, typename std::enable_if<Iter::is_hetero::value, void>::type>
      : std::true_type
    { };
  }
  template <typename T/*,*/
  /*sycl::access::mode Mode = sycl::access::mode::read_write*//*,
    typename Allocator = sycl::buffer_allocator*/>
  class device_iterator {
  protected:
    sycl::buffer<T, 1/*, Allocator*/> buffer;
    std::size_t idx;

  public:
    using value_type = T;
    using difference_type = std::make_signed<std::size_t>::type;
    using pointer = T*;
    using reference = T&;
    using iterator_category = std::random_access_iterator_tag;
    using is_hetero = std::true_type;                    // required
    static constexpr auto /*sycl::access::mode*/ mode =
      cl::sycl::access::mode::read_write; // Mode; // required

    device_iterator() : buffer(cl::sycl::range<1>(1)), idx(0) {}
    device_iterator(sycl::buffer<T, 1/*, Allocator*/> vec, std::size_t index) : buffer(vec), idx(index) {}
    device_iterator(const std::size_t n)
      : buffer(sycl::buffer<T, 1/*, Allocator*/>(sycl::range<1>(n))),
	idx(std::size_t{}) {}
                  #if 0
    reference operator*() const
    {
      auto ptr = (const_cast<device_iterator*>(this)->buffer.template get_access<mode>()).get_pointer();
      return *(ptr + idx);
    }
            #endif
            device_iterator
	    operator+(difference_type forward) const {
	      return device_iterator{buffer, idx + forward};
	    }
    difference_type operator-(const device_iterator &it) const {
      return idx - it.idx;
    }
    bool operator==( device_iterator &it) const {
      return *this - it == 0;
    }
    bool operator!=( device_iterator &it) const {
      return *this - it != 0;
    }
    template <typename OtherIterator>
    typename std::enable_if<
      internal::is_hetero_iterator<OtherIterator>::value,
      difference_type>::type
    operator-(const OtherIterator &it) const {
      return idx - std::distance(dpstd::begin(buffer), it);
    }
    std::size_t get_idx() { return idx; } // required

    sycl::buffer<T, 1/*, Allocator*/> get_buffer() { return buffer; } // required
  };

  template<class InputIter1, class InputIter2>
  void stable_sort_by_key(InputIter1 keys_first, InputIter1 keys_last, InputIter2 values_first)
  {
    auto beg1 = dpstd::zip_iterator<dpct::device_iterator<int>,
				    dpct::device_iterator<int>>(
								keys_first, values_first);

    auto beg2 = dpstd::make_zip_iterator<dpct::device_iterator<int>,
					 dpct::device_iterator<int>>(
								     keys_first, values_first);
    static_assert(std::is_same<decltype(beg1), decltype(beg2)>::value, "");
  }
}

int main() {
  const int numsH = 100;
  dpct::device_iterator<int> mapspkeyT(numsH);
  dpct::device_iterator<int> mapspvalT(numsH);
  cl::sycl::queue q;
  q.submit([&](cl::sycl::handler &h) {
      h.single_task<dpstd::zip_iterator<dpct::device_iterator<int>,
					dpct::device_iterator<int>>>([=] {});
    });

  dpct::stable_sort_by_key(mapspkeyT, mapspkeyT + numsH, mapspvalT);

  std::cout << "done.\n";
  return 0;
}

It is when I looked at the AST that I realized that it was thinking were different:
auto beg1 = dpstd::zip_iterator<dpct::device_iterator<int>,
                                                    dpct::device_iterator<int>>(
                                                                                                keys_first, keys_first);
 
|       |     `-CXXConstructExpr 0x1491f71709b0 <line:47:17, line:49:31> 'dpstd::zip_iterator< ::dpct::device_iterator<int>, ::dpct::device_iterator<int>>':'dpstd::zip_iterator< ::dpct::device_iterator<int>, ::dpct::device_iterator<int>>' 'void (const dpstd::zip_iterator< ::dpct::device_iterator<int>, ::dpct::device_iterator<int>> &)' elidable
                        
|       |           `-CXXBindTemporaryExpr 0x1491f71708d0 <line:47:17, line:49:31> 'dpstd::zip_iterator<dpct::device_iterator<int>, dpct::device_iterator<int>>':'dpstd::zip_iterator< ::dpct::device_iterator<int>, ::dpct::device_iterator<int>>' (CXXTemporary 0x1491f71708d0)
 
    auto beg2 = dpstd::make_zip_iterator<dpct::device_iterator<int>,
                                                             dpct::device_iterator<int>>(
                                                                                                     keys_first, keys_first);
|       |     `-CXXConstructExpr 0x1491f7174e80 <line:51:17, line:53:36> 'dpstd::zip_iterator<dpct::device_iterator<int>, dpct::device_iterator<int>>':'dpstd::zip_iterator<dpct::device_iterator<int>, dpct::device_iterator<int>>' 'void (const dpstd::zip_iterator<dpct::device_iterator<int>, dpct::device_iterator<int>> &)' elidable
 
|       |           `-CXXBindTemporaryExpr 0x1491f7174d70 <line:51:17, line:53:36> 'zip_iterator<dpct::device_iterator<int>, dpct::device_iterator<int>>':'dpstd::zip_iterator<dpct::device_iterator<int>, dpct::device_iterator<int>>' (CXXTemporary 0x1491f7174d70) 

beg1 was being using ::dpct::device_iterator<int> but beg2 with dpct::device_iterator<int>

With this information and after talking to Erich, the suggestion was that the getFullyQualifiedType was not the right function to get the type.

What folder should I be using for the test case?

Reply:
Alright, it feels better now, thanks for the explanation.
Can we some kind of this explanation into the test comments?
I think clang/test/SemaSYCL folder will be better for this test.